org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/JavaModelEventProvider.java

```diff
@@ -0,0 +1,106 @@
+ */
+package org.eclipse.che.jdt.ls.extension.core.internal;
+
+import static org.eclipse.che.jdt.ls.extension.api.Commands.CLIENT_UPDATE_PROJECTS_CLASSPATH;
+
+import java.util.HashSet;
+import java.util.Set;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.jdt.core.ElementChangedEvent;
+import org.eclipse.jdt.core.IElementChangedListener;
+import org.eclipse.jdt.core.IJavaElementDelta;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ResourceUtils;
+import org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer;
+
+/**
+ * Provides Java Model Element Change events to the clients
+ *
+ * @author Victor Rubezhny
+ */
+@SuppressWarnings("restriction")
+public class JavaModelEventProvider implements IElementChangedListener {
+
+  private static final int CLASSPATH_CHANGED_MASK =
+      IJavaElementDelta.F_ADDED_TO_CLASSPATH
+          | IJavaElementDelta.F_CLASSPATH_CHANGED
+          | IJavaElementDelta.F_REORDER
+          | IJavaElementDelta.F_REMOVED_FROM_CLASSPATH
+          | IJavaElementDelta.F_RESOLVED_CLASSPATH_CHANGED
+          | IJavaElementDelta.F_ARCHIVE_CONTENT_CHANGED;
+
+  private static JavaModelEventProvider INSTANCE;
+
+  private boolean isListening = false;
+
+  private JavaModelEventProvider() {}
+
+  public static JavaModelEventProvider getInstance() {
+    if (INSTANCE == null) {
+      INSTANCE = new JavaModelEventProvider();
+    }
+    return INSTANCE;
+  }
+
+  public void start() {
+    if (!this.isListening) {
+      this.isListening = true;
+      JavaCore.addElementChangedListener(this);
+    }
+  }
+
+  public void stop() {
+    if (this.isListening) {
+      this.isListening = false;
+      JavaCore.removeElementChangedListener(this);
+    }
+  }
+
+  @Override
+  public void elementChanged(ElementChangedEvent event) {
+    Set<IProject> projects = getAffectedProjects(event.getDelta(), new HashSet<IProject>());
+    if (projects.isEmpty()) {
+      return;
+    }
+
+    try {
+      Set<String> projectUris = new HashSet<String>();
+      for (IProject project : projects) {
+        projectUris.add(ResourceUtils.fixURI(project.getRawLocationURI()));
+      }
+
+      JDTLanguageServer ls = JavaLanguageServerPlugin.getInstance().getProtocol();
+      ((JDTLanguageServer) ls)
+          .getClientConnection()
+          .executeClientCommand(
+              CLIENT_UPDATE_PROJECTS_CLASSPATH,
+              (Object[]) projectUris.toArray(new String[projectUris.size()]));
+    } catch (Exception e) {
+      // Ignore.
+      JavaLanguageServerPlugin.logException(
+          "An exception occured while reporting project CLASSPATH change", e);
+    }
+  }
+
+  private Set<IProject> getAffectedProjects(
+      IJavaElementDelta delta, Set<IProject> affectedProjects) {
+    if ((delta.getFlags() & CLASSPATH_CHANGED_MASK) != 0) {
+      affectedProjects.add(delta.getElement().getJavaProject().getProject());
```

Comment:
I think `delta.getElement().getJavaProject()` may return null. It's better to check this situation.